server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/event/PopulateEmailQueryViewListener.java

```diff
@@ -0,0 +1,128 @@
+    private final EmailQueryView view;
+    private final SessionProvider sessionProvider;
+
+    @Inject
+    public PopulateEmailQueryViewListener(MessageIdManager messageIdManager, EmailQueryView view, SessionProvider sessionProvider) {
+        this.messageIdManager = messageIdManager;
+        this.view = view;
+        this.sessionProvider = sessionProvider;
+    }
+
+    @Override
+    public Group getDefaultGroup() {
+        return GROUP;
+    }
+
+    @Override
+    public boolean isHandling(Event event) {
+        return event instanceof Added
+            || event instanceof Expunged
+            || event instanceof MailboxDeletion;
+    }
+
+    @Override
+    public Publisher<Void> reactiveEvent(Event event) {
+        if (event instanceof Added) {
+            return handleAdded((Added) event);
+        }
+        if (event instanceof Expunged) {
+            return handleExpunged((Expunged) event);
+        }
+        if (event instanceof MailboxDeletion) {
+            return handleMailboxDeletion((MailboxDeletion) event);
+        }
+        return Mono.empty();
+    }
+
+    public Publisher<Void> handleMailboxDeletion(MailboxDeletion mailboxDeletion) {
+        return view.delete(mailboxDeletion.getMailboxId());
+    }
+
+    public Publisher<Void> handleExpunged(Expunged expunged) {
+        return Flux.fromStream(expunged.getUids().stream()
+            .map(uid -> expunged.getMetaData(uid).getMessageId()))
+            .concatMap(messageId -> view.delete(expunged.getMailboxId(), messageId))
+            .then();
+    }
+
+    public Mono<Void> handleAdded(Added added) {
+        MailboxSession session = sessionProvider.createSystemSession(added.getUsername());
+        return Flux.fromStream(added.getUids().stream()
+            .map(added::getMetaData))
+            .flatMap(messageMetaData -> handleAdded(added, messageMetaData, session), CONCURRENCY)
+            .then();
+    }
+
+    public Mono<Void> handleAdded(Added added, MessageMetaData messageMetaData, MailboxSession session) {
+        MessageId messageId = messageMetaData.getMessageId();
+        ZonedDateTime receivedAt = ZonedDateTime.ofInstant(messageMetaData.getInternalDate().toInstant(), ZoneOffset.UTC);
+
+        Mono<ZonedDateTime> sentAtMono = Flux.from(messageIdManager.getMessagesReactive(ImmutableList.of(messageId), FetchGroup.HEADERS, session))
+            .next()
+            .map(Throwing.function(messageResult -> Message.Builder
+                .of()
+                .use(MimeConfig.PERMISSIVE)
+                .parse(messageResult.getFullContent().getInputStream())
+                .build()))
+            .map(message -> Optional.ofNullable(message.getDate()).orElse(messageMetaData.getInternalDate()))
+            .map(date -> ZonedDateTime.ofInstant(date.toInstant(), ZoneOffset.UTC));
+
+        return sentAtMono.flatMap(sentAt -> view.save(added.getMailboxId(), sentAt, receivedAt, messageId));
```

Comment:
why not doing directly that flatmap on the `sentAtMono` above, allowing you to get rid of that intermediate variable?